File: docs/examples/java/BatchOperations.java
import java.util.*;

import com.chain.api.*;
import com.chain.exception.*;
import com.chain.http.*;
import com.chain.signing.*;

class BatchOperations {
  public static void main(String[] args) throws Exception {
    Client client = new Client();

    MockHsm.Key key = MockHsm.Key.create(client);
    HsmSigner.addKey(key, MockHsm.getSignerClient(client));

    // snippet asset-builders
    List<Asset.Builder> assetBuilders = Arrays.asList(
      new Asset.Builder()
        .setAlias("gold")
        .addRootXpub(key.xpub)
        .setQuorum(1),
      new Asset.Builder()
        .setAlias("silver")
        .addRootXpub(key.xpub)
        .setQuorum(1),
      new Asset.Builder()
        .setAlias("bronze")
        .addRootXpub(key.xpub)
        .setQuorum(0)
    );
    // endsnippet

    // snippet asset-create-batch
    BatchResponse<Asset> assetBatch = Asset.createBatch(client, assetBuilders);
    // endsnippet

    // snippet asset-create-handle-errors
    for (int i = 0; i < assetBatch.size(); i++) {
      if (assetBatch.isError(i)) {
        APIException error = assetBatch.errorsByIndex().get(i);
        System.out.println("asset " + i + " error: " + error);
      } else {
        Asset asset = assetBatch.successesByIndex().get(i);
        System.out.println("asset " + i + " created, ID: " + asset.id);
      }
    }
    // endsnippet

    // snippet nondeterministic-errors
    assetBuilders = Arrays.asList(
      new Asset.Builder()
        .setAlias("platinum")
        .addRootXpub(key.xpub)
        .setQuorum(1),
      new Asset.Builder()
        .setAlias("platinum")
        .addRootXpub(key.xpub)
        .setQuorum(1),
      new Asset.Builder()
        .setAlias("platinum")
        .addRootXpub(key.xpub)
        .setQuorum(1)
    );
    // endsnippet

    assetBatch = Asset.createBatch(client, assetBuilders);

    for (int i = 0; i < assetBatch.size(); i++) {
      if (assetBatch.isError(i)) {
        APIException error = assetBatch.errorsByIndex().get(i);
        System.out.println("asset " + i + " error: " + error);
      } else {
        Asset asset = assetBatch.successesByIndex().get(i);
        System.out.println("asset " + i + " created, ID: " + asset.id);
      }
    }

    new Account.Builder()
      .setAlias("alice")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(client);

    new Account.Builder()
      .setAlias("bob")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(client);

    // snippet batch-build-builders
    List<Transaction.Builder> txBuilders = Arrays.asList(
      new Transaction.Builder()
        .addAction(new Transaction.Action.Issue()
          .setAssetAlias("gold")
          .setAmount(100)
        ).addAction(new Transaction.Action.ControlWithAccount()
          .setAccountAlias("alice")
          .setAssetAlias("gold")
          .setAmount(100)
        ),
      new Transaction.Builder()
        .addAction(new Transaction.Action.Issue()
          .setAssetAlias("not-a-real-asset")
          .setAmount(100)
        ).addAction(new Transaction.Action.ControlWithAccount()
          .setAccountAlias("alice")
          .setAssetAlias("not-a-real-asset")
          .setAmount(100)
        ),
      new Transaction.Builder()
        .addAction(new Transaction.Action.Issue()
          .setAssetAlias("silver")
          .setAmount(100)
        ).addAction(new Transaction.Action.ControlWithAccount()
          .setAccountAlias("alice")
          .setAssetAlias("silver")
          .setAmount(100)
        )
    );
    // endsnippet

    // snippet batch-build-handle-errors
    BatchResponse<Transaction.Template> buildTxBatch = Transaction.buildBatch(client, txBuilders);

    for(Map.Entry<Integer, APIException> err : buildTxBatch.errorsByIndex().entrySet()) {
      System.out.println("Error building transaction " + err.getKey() + ": " + err.getValue());
    }
    // endsnippet

    // snippet batch-sign
    BatchResponse<Transaction.Template> signTxBatch = HsmSigner.signBatch(buildTxBatch.successes());

    for(Map.Entry<Integer, APIException> err : signTxBatch.errorsByIndex().entrySet()) {
      System.out.println("Error signing transaction " + err.getKey() + ": " + err.getValue());
    }
    // endsnippet

    // snippet batch-submit
    BatchResponse<Transaction.SubmitResponse> submitTxBatch = Transaction.submitBatch(client, signTxBatch.successes());

    for(Map.Entry<Integer, APIException> err : submitTxBatch.errorsByIndex().entrySet()) {
      System.out.println("Error submitting transaction " + err.getKey() + ": " + err.getValue());
    }

    for(Map.Entry<Integer, Transaction.SubmitResponse> success : submitTxBatch.successesByIndex().entrySet()) {
      System.out.println("Transaction " + success.getKey() + " submitted, ID: " + success.getValue().id);
    }
    // endsnippet
  }
}

File: docs/1.0/core/examples/ruby/batch_operations.rb
require 'chain'

chain = Chain::Client.new
signer = Chain::HSMSigner.new

key = chain.mock_hsm.keys.create
signer.add_key(key, chain.mock_hsm.signer_conn)

# snippet asset-builders
assets_to_build = [{
  alias: 'gold',
  root_xpubs: [key.xpub],
  quorum: 1,
}, {
  alias: 'silver',
  root_xpubs: [key.xpub],
  quorum: 1,
}, {
  alias: 'bronze',
  root_xpubs: [key.xpub],
  quorum: 0,
}]
# endsnippet

# snippet asset-create-batch
asset_batch = chain.assets.create_batch(assets_to_build)
# endsnippet

# snippet asset-create-handle-errors
asset_batch.errors.each do |index, err|
  puts "asset #{index} error: #{err}"
end

asset_batch.successes.each do |index, asset|
  puts "asset #{index} created, ID: #{asset.id}"
end
# endsnippet

# snippet nondeterministic-errors
assets_to_build = [{
  alias: 'platinum',
  root_xpubs: [key.xpub],
  quorum: 1,
}, {
  alias: 'platinum',
  root_xpubs: [key.xpub],
  quorum: 1,
}, {
  alias: 'platinum',
  root_xpubs: [key.xpub],
  quorum: 1,
}]
# endsnippet

asset_batch = chain.assets.create_batch(assets_to_build)

asset_batch.errors.each do |index, err|
  puts "asset #{index} error: #{err}"
end

asset_batch.successes.each do |index, asset|
  puts "asset #{index} created, ID: #{asset.id}"
end

# Some setup for the next examples.

chain.accounts.create(alias: 'alice', root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'bob', root_xpubs: [key.xpub], quorum: 1)

# snippet batch-build-builders
transactions_to_build = []

transactions_to_build << Chain::Transaction::Builder.new do |b|
  b.issue asset_alias: 'gold', amount: 100
  b.control_with_account account_alias: 'alice', asset_alias: 'gold', amount: 100
end

transactions_to_build << Chain::Transaction::Builder.new do |b|
  b.issue asset_alias: 'not-a-real-asset', amount: 100
  b.control_with_account account_alias: 'alice', asset_alias: 'not-a-real-asset', amount: 100
end

transactions_to_build << Chain::Transaction::Builder.new do |b|
  b.issue asset_alias: 'silver', amount: 100
  b.control_with_account account_alias: 'alice', asset_alias: 'silver', amount: 100
end
# endsnippet

# snippet batch-build-handle-errors
build_batch = chain.transactions.build_batch(transactions_to_build)

build_batch.errors.each do |index, err|
  puts "Error building transaction #{index}: #{err}"
end
# endsnippet

# snippet batch-sign
transactions_to_sign = build_batch.successes.values
sign_batch = signer.sign_batch(transactions_to_sign)

sign_batch.errors.each do |index, err|
  puts "Error signing transaction #{index}: #{err}"
end
# endsnippet

# snippet batch-submit
transactions_to_submit = sign_batch.successes.values
submit_batch = chain.transactions.submit_batch(transactions_to_submit)

submit_batch.errors.each do |index, err|
  puts "Error submitting transaction #{index}: #{err}"
end

submit_batch.successes.each do |index, submission|
  puts "Transaction #{index} submitted, ID: #{submission.id}"
end
# endsnippet
